feat: create domain entities

zrock-dev opened the pull request:
<!--- Provide a general summary of the pull request in the Title above -->
<!-- Please complete this template before creating the pull request. -->

<!-- Pull Request Template -->
#### Resolved task/US: ([CS-001 Information Receivers](https://app.clickup.com/t/868at94kb))
* * * *

## What's in this pull request?
<!-- Please include a summary of the change and which US or Task is fixed. -->
<!-- Please also include relevant motivation and context. List any dependencies that are required for this change. -->
An implementation of domain entities to develop the instrumenter

![image](https://github.com/user-attachments/assets/b9fd8ecc-005d-4e3e-84b1-15136e69268e)


## How Has This Been Tested?
<!-- How was this change tested? -->
<!-- DON'T DELETE THIS SECTION! If no tests added, explain why. -->
Implemented JUNIT tests to verify the implementation works.

## Checklist:
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] I have checked my code and corrected any misspellings

zrock-dev committed:
feat: implement domain entities

An abstract representation of Class data such as class name, methods and
which lines each method has.

- Implement ClassData to hold class name and a collection of methods.
- Implement MethodData to hold method data like: method name and method
  descriptor.
- add logging dependency
- migrate Main inside lib module

SebasthianSalpozz reviewed (changes requested):
From what I see you have a different configuration than the repository setup task, please put that in order since maybe you got involved in a task that does not belong to you

zrock-dev commented:
Dear @SebasthianSalpozz I have updated the base of the PR to point to the repository set-up PR.

SebasthianSalpozz commented on lib/build.gradle.kts.orig:
```diff
@@ -0,0 +1,109 @@
+plugins {
+    java
+    `java-library`
+<<<<<<< HEAD:build.gradle.kts
```
Error of bad rebase please solve that

zrock-dev replied:
I have deleted that file.

SebasthianSalpozz commented on lib/src/main/java/edu/usb/argosinstrumentation/domain/ClassData.java:
```diff
@@ -0,0 +1,49 @@
+    private final String className;
+    private static final Logger logger = LoggerFactory.getLogger(ClassData.class);
+
+    public ClassData(String className) {
```
Is there any reason not to use lombok for the constructor?

zrock-dev replied:
That's a great idea. I have added the Lombok annotation.

SebasthianSalpozz commented on lib/src/main/java/edu/usb/argosinstrumentation/domain/CoverageData.java:
```diff
@@ -0,0 +1,14 @@
+    private final ClassData classData;
+    private final ClassData probeData;
+
+    public CoverageData(ClassData classData, ClassData probeData) {
```
Is there any reason not to use lombok for the constructor?

zrock-dev replied:
That's a great idea. I have added the Lombok annotation.

SebasthianSalpozz commented on lib/src/main/java/edu/usb/argosinstrumentation/domain/MethodData.java:
```diff
@@ -0,0 +1,30 @@
+    private final String name;
+    private final String desc;
+
+    public MethodData(String name, String desc) {
```
Is there any reason not to use lombok for the constructor?

zrock-dev replied:
That's a great idea. I have added the Lombok annotation.

SebasthianSalpozz commented on lib/src/main/java/edu/usb/argosinstrumentation/domain/MethodData.java:
```diff
@@ -0,0 +1,30 @@
+        int result = name.hashCode();
+        result = 31 * result + desc.hashCode();
+        return result;
+    }
```
Why you don't see necessary a to String here?

zrock-dev replied:
Please clarify this comment

SebasthianSalpozz replied:
I refer to this part of code that we have in the agent :3
![image](https://github.com/user-attachments/assets/d319c572-532e-49f0-8af7-f114d114bcdd)

zrock-dev replied:
I see. Yes, indeed, I have omitted that part of code. That is because, in the agent the end result is to print the coverage report to STDOUT, but in our case we have to produce an object. That's why I think we won't need this particular piece of code, other than for debugging purposes.

zrock-dev committed:
refactor: use lombok's @Builder

Domain clases are using traditional constructors. By using Builder they
can have better control.

- Add builder annotation to domain clases.
- Remove repeated test

sebastian-barra-jalafund commented on lib/src/main/java/edu/usb/argosinstrumentation/domain/ClassData.java:
```diff
@@ -0,0 +1,48 @@
+        return MethodData.builder().name(name).desc(desc).build();
+    }
+
+    public void saveMethodData(MethodData method, int line) {
```
maybe you could add a validation here if method is null in order to avoid a NullPointerException.

zrock-dev replied:
I have added the Lombok `@NonNull` annotation to that parameter. Using a validation would be troublesome because the method would throw a null pointer exception again.

SebasthianSalpozz reviewed (commented):
just to leave evidence I tell you that when I do "" I get build failed

zamoranoalicia commented on lib/src/main/java/edu/usb/argosinstrumentation/domain/ClassData.java:
```diff
@@ -0,0 +1,50 @@
+    @Builder.Default
+    private final HashMap<MethodData, SortedSet<Integer>> methods = new HashMap<>();
+    private final String className;
+    private static final Logger logger = LoggerFactory.getLogger(ClassData.class);
```
In the future please use @Self4j  https://www.slf4j.org/ for the logs, we can keep this part as it is for now

zamoranoalicia commented on lib/src/main/java/edu/usb/argosinstrumentation/domain/ClassData.java:
```diff
@@ -0,0 +1,50 @@
+            logger.atError()
+                    .log(
+                            String.format(
+                                    "Attempt to store empty.%n Method name: %s, Method desc: %s",
```
this messages should be constants

zrock-dev replied:
Does this mean every log message should be a constant?

zrock-dev replied:
Alright I will implement this changes in a new PR

zamoranoalicia commented on lib/src/main/java/edu/usb/argosinstrumentation/domain/ClassData.java:
```diff
@@ -0,0 +1,50 @@
+
+        logger.atInfo().log(
+                String.format(
+                        "Stored method%nClass: %s%nMethod: %s %s%nAt line: %d",
```
same here this should be constanst

SebasthianSalpozz reviewed (changes requested):
I see that you add a package lib maybe its better to manage all form src

zrock-dev committed:
refactor: migrated files from lib to src

zrock-dev commented:
Dear @SebasthianSalpozz following today's meeting, I have moved the src folder from lib to the root and deleted the lib package.

zamoranoalicia reviewed (approved):
LGTM
